network/quic/net.go

```diff
@@ -0,0 +1,160 @@
+package quic
+
+import (
+	"bufio"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"io"
+	"io/ioutil"
+	"log"
+	"math/big"
+	"sync"
+	"time"
+
+	h "github.com/ConsenSys/handel"
+	"github.com/ConsenSys/handel/network"
+	quic "github.com/lucas-clemente/quic-go"
+)
+
+// Network is a handel.Network implementation using QUIC as its transport layer
+type Network struct {
+	mutex          sync.RWMutex
+	listeners      []h.Listener
+	quit           bool
+	enc            network.Encoding
+	quicListener   quic.Listener
+	sessionManager sessionManager
+}
+
+const handshakeTimeout = 2000 * time.Millisecond
+
+// NewNetwork creates Nework baked by QUIC protocol
+func NewNetwork(addr string, enc network.Encoding) (*Network, error) {
+	cfg := generateTLSConfig()
+	qCfg := &quic.Config{HandshakeTimeout: handshakeTimeout} //, AcceptCookie: f}
+	listener, err := quic.ListenAddr(addr, cfg, qCfg)
+
+	if err != nil {
+		panic(err)
+	}
+	var listeners []h.Listener
+	sessManager := newSessionManager(handshakeTimeout)
+	net := Network{
+		mutex:          sync.RWMutex{},
+		listeners:      listeners,
+		quit:           false,
+		enc:            enc,
+		quicListener:   listener,
+		sessionManager: sessManager,
+	}
+
+	go net.handler()
+	return &net, nil
+}
+
+//RegisterListener registers listener for processing incoming packets
+func (quicNet *Network) RegisterListener(listener h.Listener) {
+	quicNet.mutex.Lock()
+	defer quicNet.mutex.Unlock()
+	quicNet.listeners = append(quicNet.listeners, listener)
+}
+
+// Stop stops the network
+func (quicNet *Network) Stop() {
+	quicNet.mutex.Lock()
+	defer quicNet.mutex.Unlock()
+	quicNet.quit = true
+}
+
+//Send sends a packet to supplied identities
+func (quicNet *Network) Send(identities []h.Identity, packet *h.Packet) {
+	for _, id := range identities {
+		go quicNet.send(id, packet)
+	}
+}
+
+func (quicNet *Network) send(identity h.Identity, packet *h.Packet) {
+	dialResult := quicNet.sessionManager.Dial(identity)
+
+	if dialResult.isWaiting || dialResult.err != nil {
+		return
+	}
+	stream, err := dialResult.session.OpenStream()
+
+	byteWriter := bufio.NewWriter(stream)
+	quicNet.enc.Encode(packet, byteWriter)
+
+	if err != nil {
+		panic(err)
+	}
+	byteWriter.Flush()
+	stream.Close()
```

Comment:
do we have to close the stream ?

Reply:
yes, we create new quic session per every send. In my previous approach I tried to "cash" all the sessions but I don't know how it would scale. I can make another quic-transport implementation with session cache.

Reply:
One of @nkeywal 's argument for QUIC was the 0-RTT feature of QUIC but for that, one needs to have previously communicated with the target so the diffie hellman exchange already took place. If we don't keep the session I believe we can not use the 0-RTT feature. it's not necessarily important *now* (the platform's AWS thing is more imho) but at some point it'd be nice to have yes.